File: recipes/subms-rate-limiter/java/src/main/java/com/submillisecond/recipes/ratelimit/features/TokenBucket.java
package com.submillisecond.recipes.ratelimit.features;

import java.math.BigInteger;

/**
 * Classic token bucket: capacity {@code C}, refill rate {@code R}
 * tokens per second. Tokens accumulate up to {@code C};
 * {@link #tryAcquire(long)} drains {@code n} tokens and succeeds when
 * {@code tokens >= n}.
 *
 * <p>Different shape from the base GCRA / leaky-bucket: callers can
 * drain a variable batch in a single call, and the bucket can sit at
 * full capacity through periods of inactivity. Useful when a request
 * weight varies per-call.
 *
 * <p>State is held under a single {@code synchronized} block for
 * correctness across the {@code acquire(n)} + refill path.
 *
 * <p>Byte-equivalent to the Rust sibling
 * {@code subms_rate_limiter::TokenBucket}.
 */
public final class TokenBucket {

    private static final BigInteger SCALE = BigInteger.valueOf(1_000_000_000L);

    private final long capacity;
    private final double ratePerSec;
    private final BigInteger unitsPerNsNum;
    private final BigInteger unitsPerNsDen;
    private final Clock clock;

    // Scaled tokens (1 token = 1_000_000_000 units); held in BigInteger
    // so multi-second accumulation doesn't overflow long.
    private BigInteger tokensScaled;
    private long lastNs;

    public TokenBucket(long capacity, double ratePerSec) {
        this(capacity, ratePerSec, new SystemClock());
    }

    public TokenBucket(long capacity, double ratePerSec, Clock clock) {
        this.capacity = Math.max(1L, capacity);
        this.ratePerSec = Math.max(0.0, ratePerSec);
        // units_per_ns = (rate * SCALE) / 1_000_000_000. Stored as a
        // rational; hot path stays in integer arithmetic.
        this.unitsPerNsNum = BigInteger.valueOf((long) (this.ratePerSec * 1_000_000_000.0));
        this.unitsPerNsDen = BigInteger.valueOf(1_000_000_000L);
        this.clock = clock;
        this.lastNs = clock.nowNs();
        this.tokensScaled = BigInteger.valueOf(this.capacity).multiply(SCALE);
    }

    /** Try to drain {@code n} tokens. Returns {@code true} if granted. */
    public synchronized boolean tryAcquire(long n) {
        if (n == 0L) return true;
        if (n < 0L) throw new IllegalArgumentException("n must be >= 0");
        refillLocked();
        BigInteger want = BigInteger.valueOf(n).multiply(SCALE);
        if (tokensScaled.compareTo(want) >= 0) {
            tokensScaled = tokensScaled.subtract(want);
            return true;
        }
        return false;
    }

    /** Shorthand for {@link #tryAcquire(long) tryAcquire(1)}. */
    public boolean tryAcquireOne() {
        return tryAcquire(1L);
    }

    /** Current token count (whole tokens; fractional units truncated). */
    public synchronized long available() {
        refillLocked();
        return tokensScaled.divide(SCALE).longValueExact();
    }

    public long capacity() {
        return capacity;
    }

    public double ratePerSec() {
        return ratePerSec;
    }

    private void refillLocked() {
        long now = clock.nowNs();
        long elapsed = now - lastNs;
        if (elapsed <= 0L) return;
        BigInteger add = BigInteger.valueOf(elapsed).multiply(unitsPerNsNum).divide(unitsPerNsDen);
        BigInteger capScaled = BigInteger.valueOf(capacity).multiply(SCALE);
        tokensScaled = tokensScaled.add(add).min(capScaled);
        lastNs = now;
    }
}

File: recipes/subms-rate-limiter/rust/src/features/token_bucket.rs
//! Classic token bucket: capacity `C`, refill rate `R` tokens per
//! second. Tokens accumulate up to `C`; `try_acquire(n)` drains `n`
//! tokens and succeeds when `tokens >= n`.
//!
//! Different shape from the base GCRA / leaky-bucket: callers can drain
//! a variable batch in a single call, and the bucket can sit at full
//! capacity through periods of inactivity (the base GCRA pushes its
//! `tat` forward continuously). Useful when a request weight varies
//! per-call (e.g. "this query costs 5 tokens").
//!
//! State is held under a single `Mutex` for simplicity + correctness
//! across the `acquire(n)` + refill path. Single-atomic CAS-loop is
//! awkward when the result depends on `n`; the lock adds ~20 ns per
//! call uncontended on modern hardware.

use std::sync::Mutex;

use super::clock::{Clock, SystemClock};

struct State {
    /// Tokens currently in the bucket, scaled by `1_000_000_000` to
    /// avoid losing fractional refills between calls. A "real" token is
    /// `1_000_000_000` units here.
    tokens_scaled: u128,
    /// Last clock reading consumed in the refill calc.
    last_ns: u64,
}

pub struct TokenBucket {
    capacity: u64,
    /// Refill rate, tokens per second.
    rate_per_sec: f64,
    /// Scaled units of refill per ns. `rate_per_sec * 1_000_000_000 / 1_000_000_000`
    /// = `rate_per_sec` units per second; we store the per-ns delta to
    /// avoid floats in the hot path.
    units_per_ns_num: u128,
    units_per_ns_den: u128,
    clock: Box<dyn Clock>,
    state: Mutex<State>,
}

const SCALE: u128 = 1_000_000_000;

impl TokenBucket {
    /// Build a token bucket with `capacity` tokens, refilling at
    /// `rate_per_sec`. Starts full.
    pub fn new(capacity: u64, rate_per_sec: f64) -> Self {
        Self::with_clock(capacity, rate_per_sec, Box::new(SystemClock::new()))
    }

    /// Same as `new`, but the caller supplies the clock. Used by tests
    /// to drive a deterministic `TestClock`.
    pub fn with_clock(capacity: u64, rate_per_sec: f64, clock: Box<dyn Clock>) -> Self {
        let cap = capacity.max(1);
        let rate = rate_per_sec.max(0.0);
        // units_per_ns = (rate * SCALE) / 1_000_000_000. Stored as a
        // rational so the hot path stays in integer arithmetic.
        let num = (rate * SCALE as f64) as u128;
        let den = 1_000_000_000u128;
        let now = clock.now_ns();
        Self {
            capacity: cap,
            rate_per_sec: rate,
            units_per_ns_num: num,
            units_per_ns_den: den,
            clock,
            state: Mutex::new(State {
                tokens_scaled: (cap as u128).saturating_mul(SCALE),
                last_ns: now,
            }),
        }
    }

    /// Try to drain `n` tokens. Returns `true` if granted.
    pub fn try_acquire(&self, n: u64) -> bool {
        if n == 0 {
            return true;
        }
        let mut s = self.state.lock().unwrap();
        self.refill_locked(&mut s);
        let want = (n as u128).saturating_mul(SCALE);
        if s.tokens_scaled >= want {
            s.tokens_scaled -= want;
            true
        } else {
            false
        }
    }

    /// Shorthand for `try_acquire(1)`.
    pub fn try_acquire_one(&self) -> bool {
        self.try_acquire(1)
    }

    /// Current token count (whole tokens; fractional units truncated).
    pub fn available(&self) -> u64 {
        let mut s = self.state.lock().unwrap();
        self.refill_locked(&mut s);
        (s.tokens_scaled / SCALE) as u64
    }

    pub fn capacity(&self) -> u64 {
        self.capacity
    }

    pub fn rate_per_sec(&self) -> f64 {
        self.rate_per_sec
    }

    fn refill_locked(&self, s: &mut State) {
        let now = self.clock.now_ns();
        let elapsed = now.saturating_sub(s.last_ns) as u128;
        if elapsed == 0 {
            return;
        }
        let add = elapsed.saturating_mul(self.units_per_ns_num) / self.units_per_ns_den;
        let cap_scaled = (self.capacity as u128).saturating_mul(SCALE);
        s.tokens_scaled = (s.tokens_scaled.saturating_add(add)).min(cap_scaled);
        s.last_ns = now;
    }
}

#[cfg(test)]
#[path = "token_bucket_tests.rs"]
mod tests;
